### ISSUE TITLE
Swift Package Manager's `swift test` doesn't correctly flush its output if it goes to a pipe (usually the case in CI)

### ISSUE THREAD

**Issue by weissi** [swiftlang/swift-package-manager#6592]
### Description

When a test hangs, for example

```
import XCTest
import Glibc

final class aaaaTests: XCTestCase {
    func testExample() throws {
        while true {
            sleep(1)
        }
    }
}
```

I would expect output like

```
$ swift test
Building for debugging...
Build complete! (0.12s)
Test Suite 'All tests' started at 2023-05-19 17:09:01.252
Test Suite 'debug.xctest' started at 2023-05-19 17:09:01.253
Test Suite 'aaaaTests' started at 2023-05-19 17:09:01.253
Test Case 'aaaaTests.testExample' started at 2023-05-19 17:09:01.253
[... hang ...]
```

and that is happening. However, when `swift test`'s output goes to a pipe (which now becomes _fully buffered_ and not line buffered like stdout on a terminal by default), this happens:

```
$ swift test | cat 
Building for debugging...
Build complete! (0.43s)
[... hang ...]
```

Please note that the crucial output about which test was started is _not_ present!

At first, I thought that's an XCTest issue but

```
$ .build/debug/aaaaPackageTests.xctest | cat
Test Suite 'All tests' started at 2023-05-19 17:10:54.098
Test Suite 'debug.xctest' started at 2023-05-19 17:10:54.102
Test Suite 'aaaaTests' started at 2023-05-19 17:10:54.102
Test Case 'aaaaTests.testExample' started at 2023-05-19 17:10:54.102
[... hangs ...]
```

flushed correctly.

### Expected behavior

Output from XCTest gets written into the pipe _immediately_.

### Actual behavior

Output from XCTest gets written into the pipe only after a few kB of data (which might be _never_). Presumably because of stdio buffering.

### Steps to reproduce

1. create a new Swift Package
2. create just one test case which hangs forever
3. run `swift test | cat`

### Swift Package Manager version/commit hash

Swift Package Manager - Swift 5.9.0

### Swift & OS version (output of `swift --version ; uname -a`)

```
Swift version 5.9-dev (LLVM cee72004cfe28dc, Swift 0a083b13d08c84d)
Target: aarch64-unknown-linux-gnu
Linux fed5dc69bd9c 5.15.49-linuxkit #1 SMP PREEMPT Tue Sep 13 07:51:32 UTC 2022 aarch64 aarch64 aarch64 GNU/Linux
```

**Comment by grynspan** [comment]
@MaxDesiatov Did your changes fix this?

**Comment by MaxDesiatov** [comment]
No, the related PRs are still open due to the fact that they make CI hang.

**Comment by grynspan** [comment]
Okay, I think this should have been resolved now. If you're still seeing it, let's reopen.

**Comment by grynspan** [comment]
Duplicate of #6566

**Comment by grynspan** [comment]
(We don't currently call `setvbuf()` on the pipes we create so this may well still be an issue..)
